# First MNIST Dataset practice, section 2.1

In [2]:
import numpy as np
from keras.datasets import mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

In [3]:
print("shape of train_images: {}".format(train_images.shape))
print("length of train_labels: {}".format(len(train_labels)))
print("shape of test_images: {}".format(test_images.shape))
print("length of test_labels: {}".format(len(test_labels)))
print("10 first train labels: {}".format(train_labels[0:10]))
print("10 first test labels: {}".format(test_labels[0:10]))

shape of train_images: (60000, 28, 28)
length of train_labels: 60000
shape of test_images: (10000, 28, 28)
length of test_labels: 10000
10 first train labels: [5 0 4 1 9 2 1 3 1 4]
10 first test labels: [7 2 1 0 4 1 4 9 5 9]


Make a model with 2 fully connected layers:  
    - The first layer has 512 neurals and relu activation function  
    - The second layers has 10 neurals, ending with softmax activation would create a probability distribution at the output

In [5]:
from tensorflow.keras import models
from tensorflow.keras import layers

model = models.Sequential([
    layers.Dense(512, activation = 'relu'),
    layers.Dense(10, activation = 'softmax')
])

Compile the model with:  
    - RMS Prop optimizer  
    - Cross entropy loss function  
    - The metric of accuracy for monitoring during the training process

In [6]:
model.compile(optimizer = 'rmsprop',
             loss='sparse_categorical_crossentropy',
             metrics=['accuracy'])

Preprocess the data so that it would be fixed in to the model:  
    - train_images should be converted to the shape of (60000, 28 * 28)  
    - test_images should be converted to the shape of (10000, 28 * 28)  
    - All pixel in these images have to be converted to the range of [0, 1] (float32) (originally, the pixel value is in the range of [0, 255] (int))

In [7]:
train_images = train_images.reshape(60000, 28 * 28)
train_images = train_images.astype('float32')/255
test_images = test_images.reshape(10000, 28 * 28)
test_images = test_images.astype('float32')/255

Train the model in 5 epochs and the batch size of 128

In [8]:
model.fit(train_images, train_labels, epochs = 5, batch_size = 128)

Epoch 1/5
60000/60000 [==============================] - 5s 92us/sample - loss: 0.2589 - acc: 0.9248
Epoch 2/5
60000/60000 [==============================] - 5s 88us/sample - loss: 0.1027 - acc: 0.9691
Epoch 3/5
60000/60000 [==============================] - 5s 88us/sample - loss: 0.0675 - acc: 0.9799
Epoch 4/5
60000/60000 [==============================] - 5s 87us/sample - loss: 0.0497 - acc: 0.9850
Epoch 5/5
60000/60000 [==============================] - 5s 86us/sample - loss: 0.0370 - acc: 0.9895


Lasting 5 epochs, the accuracy raises to 98.9%  
We do some test cases with the data in test_images

In [18]:
test_digits = test_images[0:10]
predictions = model.predict(test_digits)
print("probability distribution for the first digit:\n {}".format(predictions[0]))
print("prediction for test_digits[0]: {}, with the accuracy of: {:.8f}".format(predictions[0].argmax(), predictions[0][7]))
print("The label result: {}".format(test_labels[0]))

probability distribution for the first digit:
 [4.366e-09 5.390e-11 7.370e-07 3.325e-06 4.177e-12 3.317e-08 2.257e-15
 1.000e+00 1.046e-08 1.770e-07]
prediction for test_digits[0]: 7, with the accuracy of: 0.99999571
The label result: 7


The 8th value is nearly 1.0 => 7 should be the number in this picture  
Next, we evaluate the quality of the model with all image in test_images set

In [21]:
test_loss, test_acc = model.evaluate(test_images, test_labels)
print("test accuracy: {:.3f}".format(test_acc))

10000/10000 [==============================] - 1s 54us/sample - loss: 0.0661 - acc: 0.9801
test accuracy: 0.980
